### PR TITLE
Fix issue that segment index might be wrong in slice->segments

### PR DESCRIPTION
**Pull request by jingwen-yang-yjw**
### 1. Issue description
For foreign table, if its remote server specifies the option `num_segments` and the value of option `num_segments` is larger than the number of local gpdb actual segments,
accessing such foreign table might encounter an issue as follows.
```
testdb=# insert into l_tbl select * from f_tbl limit 5;
FATAL:  unexpected content id 3, should be [-1, 2]
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Succeeded.
```

We can reproduce the issue by steps below.
```
create extension postgres_fdw;
create server multi_pg_servers foreign data wrapper postgres_fdw options (mpp_execute 'all segments', num_segments '4', multi_hosts '10.117.190.206 10.117.190.206 10.117.190.206 10.117.190.206', multi_ports '5432 5555 6666 7777', dbname 'pg_testdb');
create user mapping for gpadmin server multi_pg_servers options (user 'postgres');
create foreign table f_tbl (c1 int, c2 int) server multi_pg_servers options (table_name 'demo_local_tbl');

create table l_tbl (c1 int, c2 int);

insert into l_tbl select * from f_tbl limit 10;
```

### 2. The reason why this issue might arise

https://github.com/greenplum-db/gpdb/blob/8b135bf3ffa7cec6852e79ec96f3e4f9650fb186/src/backend/optimizer/plan/createplan.c#L3106-L3110

`sendSlice->segindex` is the index of gpdb actual segments, so it can't be more than the number of segments.

```
testdb=# explain verbose insert into l_tbl select * from f_tbl limit 5;
                                            QUERY PLAN
---------------------------------------------------------------------------------------------------
 Insert on public.l_tbl  (cost=100.00..100.22 rows=2 width=8)
   ->  Redistribute Motion 1:3  (slice1; segments: 1)  (cost=100.00..100.22 rows=2 width=8)
         Output: f_tbl.c1, f_tbl.c2
         Hash Key: f_tbl.c1
         ->  Limit  (cost=100.00..100.10 rows=5 width=8)
               Output: f_tbl.c1, f_tbl.c2
               ->  Gather Motion 4:1  (slice2; segments: 4)  (cost=100.00..100.40 rows=20 width=8)
                     Output: f_tbl.c1, f_tbl.c2
                     ->  Foreign Scan on public.f_tbl  (cost=100.00..100.15 rows=5 width=8)
                           Output: f_tbl.c1, f_tbl.c2
                           Remote SQL: SELECT c1, c2 FROM public.demo_local_table LIMIT 5::bigint
 Optimizer: Postgres-based planner
(12 rows)
```

But for this SQL, `subpath->locus.numsegments` will be 4.
Then `sendSlice->segindex` might be 3 which is not in [-1, 2].
That is why this issue arrises.

So we modify the method to decide segment index to fix this issue.

Co-authored-by: Yongtao Huang <yongtaoh@vmware.com>

### CODE REVIEW

File: contrib/postgres_fdw/sql/gp_postgres_fdw.sql [hunk 1]
```diff
@@ -111,6 +111,16 @@ set search_path=postgres_fdw_gp;
 alter server loopback options(add num_segments '4');
 EXPLAIN (VERBOSE, COSTS FALSE) SELECT * FROM gp_ft1;
 EXPLAIN ANALYZE SELECT * FROM gp_ft1;
+-- test whether segment index in slice->segments is valid or not when
+-- the value of option num_segments is larger than the number of gpdb actual segments.
+CREATE TABLE tmp_tbl (
```
Comment (hyongtao-db) [inline contrib/postgres_fdw/sql/gp_postgres_fdw.sql:116]: Just a tiny suggestion, prefer to use a name that has special meaning, like `diff_no_segment` or sth else,  rather than`tmp_tbl` in the test case

**Review by hyongtao-db — APPROVED** [review]
Code change looks good to me, just a tiny suggestion.

**Review by zhaorui20005 — APPROVED** [review]

File: src/backend/optimizer/plan/createplan.c [hunk 2]
```diff
@@ -3104,10 +3104,21 @@ create_motion_plan(PlannerInfo *root, CdbMotionPath *path)
...
+				sendSlice->numsegments = 1;
+				/*
+				 * sendSlice->segindex must be smaller than the number of gpdb actual segments.
+				 *
+				 * For foreign table, subpath->locus.numsegments might be larger than the number of
+				 * gpdb actual segments.
+				 *
+				 * So we need to use the minimum of numsegments and getgpsegmentCount() here.
+				 */
+				int gp_segment_count = getgpsegmentCount();
```
Comment (kainwen) [inline src/backend/optimizer/plan/createplan.c:3118]: Var declaration better to put at the head of the `{}` block.

File: src/backend/optimizer/plan/createplan.c [hunk 3]
```diff
@@ -3116,10 +3127,21 @@ create_motion_plan(PlannerInfo *root, CdbMotionPath *path)
 			break;
 
 		case CdbLocusType_SegmentGeneral:
-			sendSlice->gangType = GANGTYPE_SINGLETON_READER;
-			sendSlice->numsegments = subpath->locus.numsegments;
-			sendSlice->segindex = gp_session_id % subpath->locus.numsegments;
-			break;
+			{
```
Comment (kainwen) [inline src/backend/optimizer/plan/createplan.c:3130]: Foreign table currently does not support replicated distribution, why the change here? Hit a crash?

Comment (jingwen-yang-yjw) [reply src/backend/optimizer/plan/createplan.c:3130]: Yes, hitting a possible crash is a reason.

And foreign table might support replicated distribution soon. 
There is a open PR about this feature, which is under review. (https://github.com/greenplum-db/gpdb/pull/16643)

So I think it's more safe to change here too.

Comment (kainwen) [reply src/backend/optimizer/plan/createplan.c:3130]: > Yes, hitting a possible crash is a reason.
> 
> And foreign table might support replicated distribution soon. There is a open PR about this feature, which is under review. (#16643)
> 
> So I think it's more safe to change here too.

Personally I am not a fan of current code depends on later not merged PRs.

The code here looks safer but we don't add a case for it.

So
* if we can reproduce crash because of the code path here, then adding a case in this PR
* if not, then let's pick the code into the later PR and add a case their as a separate commit

Comment (jingwen-yang-yjw) [reply src/backend/optimizer/plan/createplan.c:3130]: Yeah, you're right.
I think we can't reproduce crash causing by this code path now.
I'll remove these changes in this PR and add them in the later PR if necessary.

**Review by kainwen — APPROVED** [review]
